### ISSUE TITLE
ProductRepository returns wrong product if SKU matches entity ID of another product

### ISSUE THREAD

**Issue by ktruehl** [techdivision/import-product#136]
Hello again! I found another nice one.

In `TechDivision\Import\Product\Repositories\ProductRepository` in the method `findOneBySku()`, first the cache is checked to see if a product has already been loaded. The thing is that the cache is checked using the SKU as cache key, while storing product data uses the entity ID as cache key. You use references to make this possible. However, it might happen that the SKU `[sku]` of product `A` matches the entity ID of product `B`. Furthermore assume that product `B` is loaded before product `A`. In that case, if the method `findOneBySku()` is called with `[sku]` as parameter, the call `$this->resolveReference($cacheKey)` will return `[sku]`, since no reference for product `A` exists yet. But than `$this->cache[$this->resolveReference($cacheKey)]` will return the product data for `B`, which has interesting (as in horrible) side-effects. For example, the URL rewrite update observer will try to insert the URL rewrite for `A` with the SKU instead of the entity ID as `entity_pk`. And since product `B` has already been processed, a rewrite with that primary key already exists. Boom!

Now, I created a patch, which checks whether the product data returned from the cache is actually the one we want. In I replaced
```
if ($this->isCached($sku)) {
    return $this->fromCache($sku);
}
```
with
```
if ($this->isCached($sku)) {
    $product = $this->fromCache($sku);
    if ($product[MemberNames::SKU] == $sku)
        return $product;
}
```

But in general I think that the whole reference mechanism for accessing cached items is pretty dangerous, since the assumption cannot be made that there are no conflicts between actual cache keys and references.

**Comment by wagnert** [comment]
Hi, thanks for reporting👍🏻I’ll check this as soon as possible, but not before next week as i’m out of office till then. Hope this is ok! 🙂

**Comment by ktruehl** [comment]
No worries. The patch I use is simple enough and above all has no side-effects. So I'm good for now.

**Comment by wagnert** [comment]
@ktruehl Hi, i've tested it and figured out, that the referencing is not the problem in general, but resolving the references works recursive which leads to the problem. I'll check if we need the recursive resolving, if not i'll remove it with the next version.

**Comment by wagnert** [comment]
Closed with version 3.3.0 + 2.4.0.
